src/main/java/com/paloaltonetworks/osc/api/PANManagerSecurityGroupApi.java

```diff
@@ -1,110 +1,228 @@
+ */
+package com.paloaltonetworks.osc.api;
+
+import static com.paloaltonetworks.panorama.api.methods.PanoramaApiClient.makeEntryElement;
+import static com.paloaltonetworks.utils.TagToSGIdUtil.securityGroupTag;
+import static com.paloaltonetworks.utils.VsToDevGroupNameUtil.devGroupName;
+import static java.util.Collections.emptyList;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.osc.sdk.manager.api.ManagerSecurityGroupApi;
+import org.osc.sdk.manager.element.ApplianceManagerConnectorElement;
+import org.osc.sdk.manager.element.ManagerSecurityGroupElement;
+import org.osc.sdk.manager.element.SecurityGroupMemberElement;
+import org.osc.sdk.manager.element.SecurityGroupMemberListElement;
+import org.osc.sdk.manager.element.VirtualSystemElement;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.paloaltonetworks.panorama.api.mapping.AddressEntry;
+import com.paloaltonetworks.panorama.api.mapping.GetAddressResponse;
+import com.paloaltonetworks.panorama.api.mapping.GetTagResponse;
+import com.paloaltonetworks.panorama.api.methods.PanoramaApiClient;
+import com.paloaltonetworks.utils.QuickXmlizerUtil;
+
+/**
+ * This documents "Device Management Apis"
+ */
+public class PANManagerSecurityGroupApi implements ManagerSecurityGroupApi {
+
+    private static final Logger LOG = LoggerFactory.getLogger(PANManagerSecurityGroupApi.class);
+
+    public static final String SG_TAG_PREFIX = "OSC_SecurityGroup_";
+
+    private VirtualSystemElement vs;
+    private ApplianceManagerConnectorElement mc;
+    private PanoramaApiClient panClient;
+
+    private String addrXpath;
+    private String tagXpath;
+    private String devGroup;
+
+    public PANManagerSecurityGroupApi(ApplianceManagerConnectorElement mc, VirtualSystemElement vs,
+            PanoramaApiClient panClient) {
+        this.vs = vs;
+        this.mc = mc;
+        this.panClient = panClient;
+
+        this.devGroup = devGroupName(vs.getId());
+        this.addrXpath = String.format(PanoramaApiClient.XPATH_DEVGROUP_TEMPL + "/address", this.devGroup);
+        this.tagXpath = String.format(PanoramaApiClient.XPATH_DEVGROUP_TEMPL + "/tag", this.devGroup);
+    }
+
+    /**
+    *
+    * Note that the order of arguments is reversed: {@code updateSecurityGroup} has {@code name} as a first argument
+    * and {@code id} as a second.
+    *
+    * @see ManagerSecurityGroupApi
+    *
+    */
+    @Override
+    public String createSecurityGroup(String name, String sgId, SecurityGroupMemberListElement memberList)
+            throws Exception {
+        try {
+            LOG.info("Adding security group {} with members {}", name, memberList);
+            doUpdateSecurityGroup(sgId, name, memberList);
+        } catch (Exception e) {
+            LOG.error("Exception adding security group " + name, e);
+            return null;
+        }
+
+        return sgId;

```

Comment:
we should not return the sgId passed in. We should return the tag id created on panorama. So this would be of the form SG_TAG_PREFIX + sgId